zebra-chain/src/block.rs

```diff
@@ -80,3 +87,38 @@ impl<'a> From<&'a Block> for Hash {
         (&block.header).into()
     }
 }
+
+impl SafePreallocate for Hash {
+    fn max_allocation() -> u64 {
+        // A block Hash has takes 32 bytes so we can never receive more than (MAX_PROTOCOL_MESSAGE_LEN / 32) in a single message
+        (MAX_PROTOCOL_MESSAGE_LEN as u64) / 32
+    }
+}
+
+#[cfg(test)]
+mod test_safe_preallocate {
+    use super::{Hash, MAX_PROTOCOL_MESSAGE_LEN};
+    use crate::serialization::{SafePreallocate, ZcashSerialize};
+    use proptest::prelude::*;
+    use std::convert::TryInto;
+    proptest! {
+
+        #![proptest_config(ProptestConfig::with_cases(200))]
+
+
+        #[test]
+        fn block_hash_max_allocation(hash in Hash::arbitrary_with(())) {
+            let max_allocation: usize = Hash::max_allocation().try_into().unwrap();
+            let mut smallest_disallowed_vec = Vec::with_capacity(max_allocation + 1);
+            for _ in 0..(Hash::max_allocation()+1) {
+                smallest_disallowed_vec.push(hash.clone());
```

Comment:
Clippy: `clone` isn't needed, because `hash` implements `Copy`
```suggestion
                smallest_disallowed_vec.push(hash);
```